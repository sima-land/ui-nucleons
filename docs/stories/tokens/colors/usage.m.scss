@use 'pkg:@sima-land/ui-nucleons/colors';

.imports {
  display: flex;
  flex-direction: column;
  gap: 16px;
  margin-bottom: 32px;
  > * {
    width: 100%;
  }
}
