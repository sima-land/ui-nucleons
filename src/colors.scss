// GENERATED FILE - DO NOT CHANGE IT MANUALLY

// basic
$basic-blue: #1f84db;
$basic-gray87: #212121;
$basic-gray76: #3a3a3b;
$basic-gray66: #545455;
$basic-gray54: #757575;
$basic-gray38: #9e9e9e;
$basic-gray24: #c2c2c2;
$basic-gray12: #e0e0e0;
$basic-gray8: #ebebeb;
$basic-gray4: #f5f5f5;
$basic-gray2: #fafafa;
$basic-white: #fff;

// additional
$additional-deep-red: #d50000;
$additional-red: #fb3a2f;
$additional-light-red: #feebea;
$additional-dark-teal: #089176;
$additional-teal: #09ab8b;
$additional-green: #00c853;
$additional-light-green: #64dd17;
$additional-lime: #aeea00;
$additional-faded-green: #eff9ea;
$additional-pink: #e82e5c;
$additional-purple: #b52ea8;
$additional-violet: #902bd0;
$additional-deep-purple: #634bdf;
$additional-electric-blue: #2962ff;
$additional-light-blue: #0091ea;
$additional-cyan: #00b8d4;
$additional-sky: #e4f1f9;
$additional-deep-orange: #ff7200;
$additional-amber: #ffab00;
$additional-yellow: #ffd600;
$additional-gold: #d5a43b;
$additional-brown: #795548;
$additional-blue-gray: #607d8b;
$additional-deep-blue: #00599d;
$additional-dark-blue: #002b41;
$additional-unlit-blue: #1b75c2;
$additional-crimson: #f4446b;
