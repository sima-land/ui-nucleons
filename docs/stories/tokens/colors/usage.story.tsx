import { COLORS } from '@sima-land/ui-nucleons/colors';
import { useEffect, useRef, useState } from 'react';
import ClipboardJS from 'clipboard';
import { Button } from '@sima-land/ui-nucleons/button';
import { Layout } from '@sima-land/ui-nucleons/layout';
import { Input } from '@sima-land/ui-nucleons/input';
import { TextButton } from '@sima-land/ui-nucleons/text-button';
import CopySVG from '@sima-land/ui-quarks/icons/16x16/Stroked/Copy';
import CopyBigSVG from '@sima-land/ui-quarks/icons/24x24/Stroked/Copy';
import styles from './usage.m.scss';

export const meta = {
  category: 'Токены/Colors',
  title: 'Использование',
  parameters: {
    layout: 'padded',
  },
};

export default function Usage() {
  return (
    <Layout>
      <h2>Токены цветов</h2>

      <p>Библиотека предоставляет утилиты для использования цветов дизайн-системы.</p>

      <div className={styles.section}>
        <div className={styles.imports}>
          <ImportField label='Импорт SCSS' value="@use '@sima-land/ui-nucleons/colors';" />
          <ImportField
            label='Импорт JS'
            value="import { COLORS } from '@sima-land/ui-nucleons/colors';"
          />
        </div>
      </div>

      <div className={styles.section}>
        <div className={styles.colors}>
          {[...COLORS.entries()].map(([name, value], index) => (
            <ColorView key={index} {...{ name, value }} />
          ))}
        </div>
      </div>
    </Layout>
  );
}

function ImportField({ value, label }: { value: string; label: string }) {
  const inputRef = useRef<HTMLInputElement>(null);
  const buttonRef = useRef<HTMLButtonElement>(null);
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    if (buttonRef.current) {
      const clipboard = new ClipboardJS(buttonRef.current, { text: () => value });

      clipboard.on('success', () => {
        setCopied(true);
        setTimeout(() => setCopied(false), 2000);
      });

      return () => clipboard.destroy();
    }
  }, []);

  return (
    <Input
      label={label}
      value={copied ? 'Скопировано' : value}
      disabled={copied}
      inputRef={inputRef}
      readOnly
      adornmentEnd={
        <TextButton
          disabled={copied}
          color='basic-gray87'
          buttonRef={buttonRef}
          onMouseDown={e => e.preventDefault()}
        >
          <CopyBigSVG fill='currentColor' />
        </TextButton>
      }
    />
  );
}

function ColorView({ name, value }: { name: string; value: string }) {
  return (
    <div className={styles.color}>
      <span className={styles.circle} style={{ background: value }} />
      <span className={styles.title}>{name}</span>
      <span>{value}</span>

      <CopyButton title='SCSS' value={`colors.$${name}`} />
      <CopyButton title='JS' value={`COLORS.get('${name}')`} />
    </div>
  );
}

function CopyButton({ title, value }: { title: string; value: string }) {
  const buttonRef = useRef<HTMLButtonElement>(null);
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    if (buttonRef.current) {
      const clipboard = new ClipboardJS(buttonRef.current, {
        text: () => value,
      });

      clipboard.on('success', () => {
        setCopied(true);
        setTimeout(() => setCopied(false), 2000);
      });

      return () => clipboard.destroy();
    }
  }, []);

  return (
    <div>
      <Button
        ref={buttonRef}
        size='xs'
        viewType='secondary'
        icon={!copied ? CopySVG : undefined}
        disabled={copied}
      >
        {copied ? 'Скопировано' : title}
      </Button>
    </div>
  );
}
