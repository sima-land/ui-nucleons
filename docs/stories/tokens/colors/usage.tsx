import { COLORS } from '@sima-land/ui-nucleons/colors';
import { CardGrid, ImportField, TokenCard } from '../gradients/usage';
import styles from './usage.m.scss';

export function ColorsPicker() {
  return (
    <>
      <h2>Цвета</h2>

      <p>Библиотека предоставляет утилиты для использования цветов дизайн-системы.</p>

      <div className={styles.imports}>
        <ImportField label='Импорт SCSS' value="@use 'pkg:@sima-land/ui-nucleons/colors';" />
        <ImportField
          label='Импорт JS'
          value="import { COLORS } from '@sima-land/ui-nucleons/colors';"
        />
      </div>

      <CardGrid>
        {[...COLORS.entries()].map(([name, value], index) => (
          <TokenCard
            key={index}
            name={name}
            value={value}
            usageToken={`color/${name}`}
            usageScss={`colors.$${name}`}
            usageJs={`COLORS.get('${name}')`}
          />
        ))}
      </CardGrid>
    </>
  );
}
