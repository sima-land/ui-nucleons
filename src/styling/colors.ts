import { ClassGetter } from './utils';
import { COLORS, Token } from '../colors';
import classes from './colors.module.scss';

const isColorKey = COLORS.has.bind(COLORS);

/** @deprecated */
export const color = ClassGetter<Token>(classes, isColorKey, 'color__');

/** @deprecated */
export const hoverColor = ClassGetter<Token>(classes, isColorKey, 'hover-color__');

/** @deprecated */
export const bgColor = ClassGetter<Token>(classes, isColorKey, 'bg-color__');

/** @deprecated */
export const stroke = ClassGetter<Token>(classes, isColorKey, 'stroke__');

/** @deprecated */
export const fill = ClassGetter<Token>(classes, isColorKey, 'fill__');
