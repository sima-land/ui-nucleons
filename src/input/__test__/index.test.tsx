import { createRef, MouseEvent } from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Input } from '..';
import PlaceholderSVG from '@sima-land/ui-quarks/icons/16x16/Stroked/Placeholder';
import { COLORS } from '../../colors';

describe('Input', () => {
  it('should render default value', () => {
    const { getByTestId } = render(<Input defaultValue='Hello' />);

    expect(getByTestId('base-input:field').getAttribute('value')).toBe('Hello');
  });

  it('should render value', () => {
    const { getByTestId } = render(<Input value='World' onChange={jest.fn()} />);

    expect(getByTestId('base-input:field').getAttribute('value')).toBe('World');
  });

  it('should render caption', () => {
    const { getByTestId } = render(<Input caption='Some caption here' />);

    expect(getByTestId('field-block:caption').textContent).toBe('Some caption here');
  });

  it('should render start adornment', () => {
    const { queryAllByTestId } = render(
      <Input adornmentStart={<PlaceholderSVG data-testid='icon-start' />} />,
    );

    expect(queryAllByTestId('field-block:adornment-start')).toHaveLength(1);
    expect(queryAllByTestId('icon-start')).toHaveLength(1);
  });

  it('should render end adornment', () => {
    const { queryAllByTestId } = render(
      <Input adornmentEnd={<PlaceholderSVG data-testid='icon-end' />} />,
    );

    expect(queryAllByTestId('field-block:adornment-end')).toHaveLength(1);
    expect(queryAllByTestId('icon-end')).toHaveLength(1);
  });

  it('should handle "inputRef" prop', () => {
    const ref = createRef<HTMLInputElement>();

    expect(ref.current).toBe(null);
    render(<Input inputRef={ref} />);
    expect(ref.current instanceof HTMLInputElement).toBe(true);
  });

  it('should handle "baseInputProps.inputRef" prop', () => {
    const ref = createRef<HTMLInputElement>();

    expect(ref.current).toBe(null);
    render(<Input baseInputProps={{ inputRef: ref }} />);
    expect(ref.current instanceof HTMLInputElement).toBe(true);
  });

  it('should handle "buttonProps.onClick" prop', () => {
    const spy = jest.fn();

    const { getByTestId } = render(<Input blockProps={{ onClick: spy }} />);

    expect(spy).toHaveBeenCalledTimes(0);
    fireEvent.click(getByTestId('field-block:block'));
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should focus input when mousedown event fired on block', () => {
    const inputRef = createRef<HTMLInputElement>();
    const focusSpy = jest.fn();
    const mouseDownSpy = jest.fn();

    const { getByTestId } = render(
      <Input inputRef={inputRef} blockProps={{ onMouseDown: mouseDownSpy }} />,
    );

    inputRef.current?.addEventListener('focus', focusSpy);
    expect(mouseDownSpy).toHaveBeenCalledTimes(0);
    expect(focusSpy).toHaveBeenCalledTimes(0);

    fireEvent.mouseDown(getByTestId('field-block:block'));

    expect(mouseDownSpy).toHaveBeenCalledTimes(1);
    expect(focusSpy).toHaveBeenCalledTimes(1);
  });

  it('should not focus input when mousedown event fired on block and prevented', () => {
    const inputRef = createRef<HTMLInputElement>();
    const focusSpy = jest.fn();
    const mouseDownSpy = jest.fn((e: MouseEvent) => e.preventDefault());

    const { getByTestId } = render(
      <Input inputRef={inputRef} blockProps={{ onMouseDown: mouseDownSpy }} />,
    );

    inputRef.current?.addEventListener('focus', focusSpy);
    expect(mouseDownSpy).toHaveBeenCalledTimes(0);
    expect(focusSpy).toHaveBeenCalledTimes(0);

    fireEvent.mouseDown(getByTestId('field-block:block'));

    expect(mouseDownSpy).toHaveBeenCalledTimes(1);
    expect(focusSpy).toHaveBeenCalledTimes(0);
  });

  it('should properly set "--placeholder-color" variable', () => {
    const { getByTestId, rerender } = render(<Input failed disabled={false} />);

    function placeholderColor() {
      return getByTestId('base-input').style.getPropertyValue('--placeholder-color');
    }

    expect(placeholderColor()).toBe(COLORS.get('additional-deep-red'));

    rerender(<Input failed={false} disabled />);

    expect(placeholderColor()).toBe(COLORS.get('basic-gray24'));

    rerender(<Input failed={false} disabled={false} />);

    expect(placeholderColor()).toBe(COLORS.get('basic-gray38'));
  });

  it('should handle "clearable" prop', () => {
    const spy = jest.fn();
    const { queryAllByTestId, getByTestId } = render(
      <Input defaultValue='Something' clearable onClear={spy} />,
    );

    expect(queryAllByTestId('input:clear-button')).toHaveLength(0);

    fireEvent.focus(getByTestId('base-input:field'));
    expect(queryAllByTestId('input:clear-button')).toHaveLength(1);

    fireEvent.blur(getByTestId('base-input:field'));
    expect(queryAllByTestId('input:clear-button')).toHaveLength(0);

    fireEvent.focus(getByTestId('base-input:field'));
    expect(spy).toHaveBeenCalledTimes(0);
    fireEvent.click(getByTestId('input:clear-button'), { bubbles: true });
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should render clear icon depends on size', () => {
    const { rerender, getByTestId, queryAllByTestId } = render(
      <Input defaultValue='Something' clearable size='s' />,
    );
    fireEvent.focus(getByTestId('base-input:field'));
    expect(queryAllByTestId('input:clear-button')).toHaveLength(1);
    expect(getByTestId('input:clear-button')).toMatchSnapshot();

    rerender(<Input defaultValue='Something' clearable size='m' />);
    expect(queryAllByTestId('input:clear-button')).toHaveLength(1);
    expect(getByTestId('input:clear-button')).toMatchSnapshot();
  });

  it('should update filled state when "input" event fired on input element', () => {
    const spy = jest.fn();

    const { getByTestId } = render(<Input onInput={spy} />);

    expect(spy).toHaveBeenCalledTimes(0);
    fireEvent.input(getByTestId('base-input:field'), { target: { value: 'New value' } });
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should update filled state when "change" event fired on input element', () => {
    const spy = jest.fn();

    const { getByTestId } = render(<Input onChange={spy} />);

    expect(spy).toHaveBeenCalledTimes(0);
    fireEvent.change(getByTestId('base-input:field'), { target: { value: 'New value' } });
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should properly set "data-testid" to elements', () => {
    const { container, rerender } = render(<Input />);
    const find = (s: string) => container.querySelectorAll(s);

    expect(find('[data-testid="input"] input')).toHaveLength(1);
    expect(find('[data-testid="input"] [data-testid="base-input"] input')).toHaveLength(1);
    expect(find('[data-testid="input"] [data-testid="base-input:field"]')).toHaveLength(1);

    rerender(<Input data-testid='email' />);
    expect(find('[data-testid="email"] input')).toHaveLength(1);
    expect(find('[data-testid="email"] [data-testid="base-input"] input')).toHaveLength(1);
    expect(find('[data-testid="email"] [data-testid="base-input:field"]')).toHaveLength(1);
  });

  it('should handle "focused" prop', () => {
    const { getByTestId, rerender } = render(<Input />);

    expect(getByTestId('input').classList.contains('focused')).toBe(false);
    fireEvent.focus(getByTestId('base-input:field'));
    expect(getByTestId('input').classList.contains('focused')).toBe(true);
    fireEvent.blur(getByTestId('base-input:field'));
    expect(getByTestId('input').classList.contains('focused')).toBe(false);

    rerender(<Input focused />);
    expect(getByTestId('input').classList.contains('focused')).toBe(true);
  });
});
