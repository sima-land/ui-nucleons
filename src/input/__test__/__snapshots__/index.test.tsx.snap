// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render clear icon depends on size 1`] = `
<svg
  aria-label="Очистить поле"
  class="clear-button"
  data-testid="input:clear-button"
  height="16"
  role="button"
  viewBox="0 0 16 16"
  width="16"
>
  <path
    clip-rule="evenodd"
    d="M8 15A7 7 0 1 0 8 1a7 7 0 0 0 0 14Zm2.828-9.828a1 1 0 0 1 0 1.414L9.414 8l1.414 1.414a1 1 0 0 1-1.414 1.414L8 9.414l-1.414 1.414a1 1 0 1 1-1.414-1.414L6.586 8 5.172 6.586a1 1 0 0 1 1.414-1.414L8 6.586l1.414-1.414a1 1 0 0 1 1.414 0Z"
    fill-rule="evenodd"
  />
</svg>
`;

exports[`Input should render clear icon depends on size 2`] = `
<svg
  aria-label="Очистить поле"
  class="clear-button"
  data-testid="input:clear-button"
  height="24"
  role="button"
  viewBox="0 0 24 24"
  width="24"
>
  <path
    clip-rule="evenodd"
    d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10Zm4.95-6.464a1 1 0 1 1-1.414 1.414L12 13.414 8.464 16.95a1 1 0 1 1-1.414-1.414L10.586 12 7.05 8.464A1 1 0 0 1 8.464 7.05L12 10.586l3.536-3.536a1 1 0 1 1 1.414 1.414L13.414 12l3.536 3.536Z"
    fill-rule="evenodd"
  />
</svg>
`;
